observatory-dags/observatory/dags/model.py

```diff
@@ -0,0 +1,1290 @@
+    institutional repository.
+    """
+
+    oa: bool = None
+    green: bool = None
+    gold: bool = None
+    gold_doaj: bool = None
+    hybrid: bool = None
+    bronze: bool = None
+    green_only: bool = None
+
+
+@dataclass
+class Author:
+    """ An author.
+
+    :param id: unique identifier.
+    :param name: the name of the author.
+    :param institution: the author's institution.
+    """
+
+    id: int
+    name: str = None
+    institution: Institution = None
+
+
+@dataclass
+class Funder:
+    """ A research funder.
+
+    :param id: unique identifier.
+    :param name: the name of the funder.
+    :param doi: the DOI of the funder.
+    :param country_code: the country code of the funder.
+    :param region: the region the funder is located in.
+    :param funding_body_type: the funding body type, see FUNDING_BODY_TYPES.
+    :param funding_body_subtype: the funding body subtype, see FUNDING_BODY_SUBTYPES.
+    """
+
+    id: int
+    name: str = None
+    doi: str = None
+    country_code: str = None
+    region: str = None
+    funding_body_type: str = None
+    funding_body_subtype: str = None
+
+
+@dataclass
+class Publisher:
+    """ A publisher.
+
+    :param id: unique identifier.
+    :param name: the name of the publisher.
+    :param doi_prefix: the publisher DOI prefix.
+    :param journals: the journals owned by the publisher.
+    """
+
+    id: int
+    name: str = None
+    doi_prefix: int = None
+    journals: List[Journal] = None
+
+
+@dataclass
+class FieldOfStudy:
+    """ A field of study.
+
+    :param id: unique identifier.
+    :param name: the field of study name.
+    :param level: the field of study level.
+    """
+
+    id: int
+    name: str = None
+    level: int = None
+
+
+@dataclass
+class Journal:
+    """ A journal
+
+    :param id: unique identifier.
+    :param name: the journal name.
+    :param name: the license that articles are published under by the journal.
+    """
+
+    id: int
+    name: str = None
+    license: str = None
+
+
+@dataclass
+class Event:
+    """ An event.
+
+    :param source: the source of the event, see EVENT_TYPES.
+    :param event_date: the date of the event.
+    """
+
+    source: str = None
+    event_date: pendulum.Pendulum = None
+
+
+@dataclass
+class ObservatoryDataset:
+    """ The generated observatory dataset.
+
+    :param institutions:
+    :param authors:
+    :param funders:
+    :param publishers:
+    :param papers:
+    :param fields_of_study:
+    """
+
+    institutions: List[Institution]
+    authors: List[Author]
+    funders: List[Funder]
+    publishers: List[Publisher]
+    papers: List[Paper]
+    fields_of_study: List[FieldOfStudy]
+
+
+def make_doi(doi_prefix: int):
+    """ Makes a randomised DOI given a DOI prefix.
+
+    :param doi_prefix: the DOI prefix.
+    :return: the DOI.
+    """
+
+    return f"10.{doi_prefix}/{str(uuid.uuid4())}"
+
+
+def make_observatory_dataset(
```

Comment:
Perhaps we can break this function up a bit.  It seems to be doing a lot of things in there.